docs/RFCS/20200720_modern_passwords.md

```diff
@@ -0,0 +1,600 @@
+
+[HBA configuration]: https://dr-knz.net/authentication-in-postgresql-and-cockroachdb.html
+
+For example, pg's authn could be implemented as follows:
+
+- StoreFn, storage `password_encryption=md5`: compute MD5 hash
+- HBA rule chooses `md5` authn method:
+  - ChallengeFn: "please send MD5 hash please"
+  - ResponseFn: compute MD5 hash of password, send MD5 hash to server
+  - ValidateFn: compare provided MD5  hash to stored MD5 hash
+
+But notice that it could *also* be implemented as follows:
+
+- StoreFn, storage `password_encryption=md5`: compute MD5 hash
+- HBA rule chooses `password` (cleartext) authn method:
+  - ChallengeFn: "please send cleartext password please"
+  - ResponseFn: send cleartext password to server
+  - ValidateFn: server computes MD5 hash of provided password server-side, then compare computed MD5  hash to stored MD5 hash
+
+Or also as follows:
+
+- StoreFn, storage `password_encryption=password`: store plaintext password
+- HBA rule chooses `md5` authn method:
+  - ChallengeFn: "please send MD5 hash please"
+  - ResponseFn: compute MD5 hash of password, send MD5 hash to server
+  - ValidateFn: server computes MD5 hash of *stored* password, then compares computed hash with provided hash.
+
+All these combinations happen to be possible.
+
+PostgreSQL supports the following combinations of storage/authn methods:
+
+| PostgreSQL `password_encryption` | Storage method                                 | Supported authentication methods |
+|----------------------------------|------------------------------------------------|----------------------------------|
+| `password`                       | Password stored as-is, in cleartext            | `password`, `md5`                |
+| `md5`, `on`,                     | MD5 hash of password stored                    | `password`, `md5`                |
+| `scram-sha-256` (since v10)      | Stores SCRAM hash, salt, iterations parameters | `password`, `scram-sha-256`      |
+
+It is thus possible for a single storage method to support multiple
+authn methods. Certain combinations are impossible because of the
+definition of the password algorithm. For example, it is impossible to
+implement the `md5` authn method if the `scram-sha-256` storage method
+was used.
+
+## CockroachDB's password methods
+
+CockroachDB currently supports a single *storage method* for password,
+derived from the BCrypt algorithm. It is separate from and incompatible
+with all the storage methods supported by PostgreSQL.
+We'll call it `bcrypt-crdb` in the rest of the RFC.
+
+It also only supports the `password` *authn method*.
+
+## Security considerations
+
+- Storage methods:
+  - `password` offers no security against attackers who get hold of stored passwords.
+  - `md5` *used* to offer some security, offers none today.
+  - `scram-sha-256` offers good security against attackers who get hold of stored password.
+  - CockroachDB's `bcrypt-crdb` storage method also offers OK security.
```

Comment:
it seems that the storage methods of scram-sha-256 and bcrypt-crdb are equivalent in security terms, i.e. one is not better than the other. I would consider using the same word to describe them either both OK or both good, otherwise it may appear that they are different in the protections they offer.

Reply:
Done.